tools/java/org/unicode/cldr/util/XMLSource.java

```diff
@@ -1747,4 +1764,711 @@ public DtdType getDtdType() {
+        private XMLSource source;
+        private String lastActiveLeafNode;
+        private String lastLeafNode;
+        private int isSupplemental = -1;
+        private int[] orderedCounter = new int[30]; // just make deep enough to handle any CLDR file.
+        private String[] orderedString = new String[30]; // just make deep enough to handle any CLDR file.
+        private int level = 0;
+        private int overrideCount = 0;
+
+        XMLNormalizingHandler(XMLSource source, DraftStatus minimalDraftStatus) {
+            this.source = source;
+            this.minimalDraftStatus = minimalDraftStatus;
+        }
+
+        private String show(Attributes attributes) {
+            if (attributes == null) return "null";
+            String result = "";
+            for (int i = 0; i < attributes.getLength(); ++i) {
+                String attribute = attributes.getQName(i);
+                String value = attributes.getValue(i);
+                result += "[@" + attribute + "=\"" + value + "\"]"; // TODO quote the value??
+            }
+            return result;
+        }
+
+        private void push(String qName, Attributes attributes) {
+            // SHOW_ALL &&
+            Log.logln(LOG_PROGRESS, "push\t" + qName + "\t" + show(attributes));
+            ++level;
+            if (!qName.equals(orderedString[level])) {
+                // orderedCounter[level] = 0;
+                orderedString[level] = qName;
+            }
+            if (lastChars.length() != 0) {
+                if (whitespace.containsAll(lastChars))
+                    lastChars = "";
+                else
+                    throw new IllegalArgumentException("Must not have mixed content: " + qName + ", "
+                        + show(attributes) + ", Content: " + lastChars);
+            }
+
+            currentFullXPath += "/" + qName;
+            if (dtdData.isOrdered(qName)) {
+                currentFullXPath += orderingAttribute();
+            }
+            if (attributes.getLength() > 0) {
+                attributeOrder.clear();
+                for (int i = 0; i < attributes.getLength(); ++i) {
+                    String attribute = attributes.getQName(i);
+                    String value = attributes.getValue(i);
+
+                    if (attribute.equals("cldrVersion")
+                        && (qName.equals("version"))) {
+                        ((SimpleXMLSource) source).setDtdVersionInfo(VersionInfo.getInstance(value));
+                    } else {
+                        putAndFixDeprecatedAttribute(qName, attribute, value);
+                    }
+                }
+                for (Iterator<String> it = attributeOrder.keySet().iterator(); it.hasNext();) {
+                    String attribute = it.next();
+                    String value = attributeOrder.get(attribute);
+                    String both = "[@" + attribute + "=\"" + value + "\"]"; // TODO quote the value??
+                    currentFullXPath += both;
```

Comment:
If there are many string concatenations in a loop consider StringBuilder

Reply:
Agreed.

Reply:
Agree and will change.